nengo_loihi/builder/inputs.py

```diff
@@ -47,10 +52,11 @@ class ChipReceiveNode(Node):
         Spike input that will inject the spikes from the sender onto the chip.
     """
 
-    def __init__(self, dimensions, size_out, label=Default):
+    def __init__(self, dimensions, size_out, label=Default, output=None):
         self.raw_dimensions = dimensions
         self.spike_target = None
-        super().__init__(self.update, size_in=0, size_out=size_out, label=label)
+        output = self.update if output is None else output
+        super().__init__(output, size_in=0, size_out=size_out, label=label)
```

Comment:
So I started with just overwriting `.update`, but there was some benefit to having it as a process that I honestly can't remember right now.

I would be OK with making this not be a subclass of `ChipReceiveNeurons`. I agree that it doesn't completely fit with the original purpose of that class. It can't just be a normal `Node`, because the splitter needs to know to put it on the chip, which is part of the reason I subclassed it in the first place. I would be okay having a general subclass of `ChipNode`, which `DVSFileChipNode` is a subclass of, and `ChipReceiveNode` and `ChipReceiveNeurons` are also subclasses of. Does that sound OK?

Reply:
Rather than adding `ChipNode`, I might just have two conditions in the splitter check instead (like `if isinstance(node, ChipReceiveNode) or isinstance(node.output, DVSFileImageProcess)`). It makes it a bit more awkward in the splitter, but that seems like less added complexity overall than adding a new class to the hierarchy.

Reply:
Could we get rid of `DVSFileChipNode` entirely in that case, and just have `DVSFileImageProcess` (combined with regular Nodes)?

Reply:
My idea behind the new class is that we might have more such nodes in the future. But maybe better to cross that bridge when we come to it.

The idea of doing it all as a process, and just checking for that as the output, and then getting rid of `DVSFileChipNode` entirely, is an interesting one. One potential downside is that it could cause a different way of doing things with files vs live DVS, since I'm not sure if we'd want to put in the work to allow the live DVS node to work as a process (or even in core Nengo at all), though it would certainly be cool if we did. 

It would also make things a bit more complicated in the builder, since right now I just have a different builder for `DVSFileChipNode` versus `ChipReceiveNode`. We currently don't have any special process builders in NengoLoihi. I guess we could do it in a general `Node` builder, and we'd know that any Nodes that get through to that point are being built on-chip, since otherwise they'd use the core Nengo build system. So it's definitely doable, and maybe not even too bad, but certainly a different way of going about things. Conceptually, I think I like having the separate Node type a little bit more, because it makes it clearer to users where the action is happening (that was my idea behind the name `DVSFileChipNode`, is that it indicates that the Node is happening on the chip, unlike normal Nodes). So if we move to the process, then we lose that. That would be my main argument for sticking with the current system.

Reply:
If it wouldn't be too much extra work to have the live DVS work as a process in NengoCore that'd definitely be cool (and generally useful across the whole ecosystem).

Another possibility is that we could have a `DVSFileChipNode` helper function (which under the hood would just be doing `Node(DVSFileImageProcess)`). That'd give the more visible distinction in the code/examples that this is a special kind of Node we're making (which I agree is useful), but allow us to keep things simpler under the hood.

Reply:
I'm not sure how much extra work it would be to have a live DVS process. I think you'd need to be running something like jAER in the background to actually interface with the DVS, and then connect to jAER in Python. I've always just used jAER to save files that I load, I've never done it live. @tcstewar might have experience with this. Anyway, it's a problem for future us.

A nice thing about having `ChipNode` is that it fits well into our pluggable builder system. So if somebody had some other spike data they want to pipe in, they can make their own `ChipNode` subclass that reads their file format and add a builder that puts the spikes in a `SpikeInput`. With processes, this wouldn't be possible, since they'd also have to add to the splitter that this type of process should be built by NengoLoihi, not Nengo core. To me, it seems like a clear way to group all special nodes that we want built by NengoLoihi.

That said, if you're really tied to the process idea, I could go for it. I like the idea of the `DVSFileChipNode` helper function in that case. It is true that the stuff in the process (as well as all the general DVS file code supporting it) is not NengoLoihi specific, so it's something that we could one day move to Nengo or NengoExtras, and the process model would fit much better with that.

So my preference would be to add the `ChipNode`, keep `DVSFileChipNode` as a subclass and hide the process as an implementation detail. Then down the road, if we add/move some of this functionality to Nengo/NengoExtras, we can do it as a process there. But as I said, I don't think it's a big deal either way.

Reply:
Part of the intention of the `Process` base class is to make things work with the pluggable builder system; I never intended for `Node` to be subclassed. I would rather it not work but I know that we programmed defensively enough that it works and the consensus didn't want to limit things for reasons like this, but I think there should not be reasons to subclass `Node`, it should be done with `Process` instead. If we had more resources for NengoLoihi, I would make `ChipNode` and other node subclasses into processes and refactor the splitter accordingly.

Reply:
So my question is what's an extensible way for users to indicate that a node should be built by the NengoLoihi build system rather than the core Nengo builder; i.e., how does a user tell the splitter that their process type should be put on the Loihi? If we can find a way to do that, then I agree that the process system is better.

Maybe we need a process subtype that the splitter automatically puts on chip? i.e. a `ChipProcess` rather than a `ChipNode`?

Or we could do something with the config system, though I feel like it's not ideal for this particular case (though maybe I'm just not able to imagine a good solution with it).